instrgen/docs/design.md

```diff
@@ -0,0 +1,200 @@
+# OpenTelemetry Go Source Automatic Instrumentation - design
+
+This document describes design of `instrgen` - compile time automatic OpenTelemetry instrumentation for Go programming language.
+
+## Background
+
+Usually techniques used for instrumenting programs in specific languages are connected to runtimes and type systems
+used by these languages and depend on capabilities offered by them. Methods differs depending on language properties.
+Are they executed by virtual machine? Are they statically or dynamically typed?
+Language and its internals influence instrumentation techniques.
+In dynamic languages it's very easy to replace original implementation. This is done by providing a new
+implementation built on top of the original one. A technique called monkey patching (i.e. Python decorators).
+
+Go is statically typed, natively compiled programming language which means that works on bare methal without
+virtual machine in between. The only component that lives between your program and machine is operating system.
+Because of this limitation, auto-instrumentation of Go programs is limited to the following:
+
+1. Auto-instrument within the kernel itself (i.e. [eBPF](https://ebpf.io/))
+2. Auto-instrument during compilation of the program
+
+The second approach is taken by `instrgen`.
+
+## Rationale
+
+Compile time instrumentation is not different by other mechanisms described above.
+The fundamental difference is the injection point, instrumentation has to be done 
+at compile time which has cons and pros.
+
+Go is perfect language for compile time instrumentation for two main reasons:
+- everything that's needed is in standard library
+- it's easy to control compilation process due to go compiler architecture
+
+## Design
```

Comment:
This document is missing an alternatives considered section.  Were there any other approaches you explored to doing compile-time instrumentation (e.g. alternatives to toolexec)?  Other than choosing not to use eBPF, were there important design decisions you made when creating this?

Reply:
I started without doing `toolexec`, however `toolexec` has this advantage that you are in the middle of compilation process, so you can see and control everything what go compiler is doing.

Reply:
I'm quite ignorant of the tools available in this space, so part of this is helping me and others understand what you could have done, but didn't decide to do.